fw/http_msg.c

```diff
@@ -1492,36 +1492,57 @@ tfw_http_msg_expand_from_pool_lc(TfwHttpTransIter *mit, TfwPool* pool,
 	return __tfw_http_msg_expand_from_pool(mit, pool, str, tfw_cstrtolower);
 }
 
-static void
-__tfw_h2_msg_shrink_frags(TfwMsgIter *it, const char *body,
-			  TfwHttpRespCleanup *cleanup)
+static inline void
+__tfw_h2_msg_move_frags(struct sk_buff *skb, int frag_idx,
+			TfwHttpRespCleanup *cleanup)
 {
-	struct skb_shared_info *si = skb_shinfo(it->skb);
-	skb_frag_t *frag = &si->frags[it->frag];
-	char* frag_addr = skb_frag_address(frag);
-	int len;
 	int i;
+	int len;
+	struct page *page;
+	struct skb_shared_info *si = skb_shinfo(skb);
 
-	/* Delete all fragments from skb. We will use this skb as skb_head. */
-	if (!body) {
-		for (i = 0; i < si->nr_frags; i++) {
-			struct page *pg = skb_frag_page(&si->frags[i]);
+	for (i = 0; i < frag_idx; i++) {
+		page = skb_frag_page(&si->frags[i]);
+		cleanup->pages[i] = compound_head(page);
+		cleanup->pages_sz++;
+		len = skb_frag_size(&si->frags[i]);
+		ss_skb_adjust_data_len(skb, -len);
```

Comment:
It'd be better to accumulate all the frag length and call only one `ss_skb_adjust_data_len()` after the loop